packages/client/src/links/wsLink/wsClient/index.ts

```diff
@@ -0,0 +1,396 @@
+  TRPCClientIncomingRequest,
+  TRPCClientOutgoingMessage,
+  TRPCResponseMessage,
+} from '@trpc/server/unstable-core-do-not-import';
+import { run, sleep } from '@trpc/server/unstable-core-do-not-import';
+import { TRPCClientError } from '../../../TRPCClientError';
+import type { TRPCConnectionState } from '../../internals/subscriptions';
+import type { Operation } from '../../types';
+import type { WebSocketClientOptions } from './options';
+import { exponentialBackoff, keepAliveDefaults, lazyDefaults } from './options';
+import {
+  ReconnectManager,
+  TRPCWebSocketReconnectFatal,
+} from './reconnectManager';
+import type { TCallbacks } from './requestManager';
+import { RequestManager } from './requestManager';
+import {
+  buildConnectionMessage,
+  prepareUrl,
+  ResettableTimeout,
+  TRPCWebSocketClosedError,
+} from './utils';
+import { backwardCompatibility, WsConnection } from './wsConnection';
+
+/**
+ * A WebSocket client for managing TRPC operations, supporting lazy initialization,
+ * reconnection, keep-alive, and request management.
+ */
+export class WsClient {
+  /**
+   * Observable tracking the current connection state, including errors.
+   */
+  public readonly connectionState: BehaviorSubject<
+    TRPCConnectionState<TRPCClientError<AnyTRPCRouter>>
+  >;
+
+  private requestManager = new RequestManager();
+  private readonly activeConnection: WsConnection;
+  private inactivityTimeout: ResettableTimeout;
+  private readonly reconnectManager: ReconnectManager;
+  private readonly callbacks: Pick<
+    WebSocketClientOptions,
+    'onOpen' | 'onClose' | 'onError'
+  >;
+  private readonly connectionParams: WebSocketClientOptions['connectionParams'];
+  private readonly lazyEnabled: boolean;
+
+  constructor(opts: WebSocketClientOptions) {
+    // Initialize callbacks, connection parameters, and options.
+    this.callbacks = {
+      onOpen: opts.onOpen,
+      onClose: opts.onClose,
+      onError: opts.onError,
+    };
+    this.connectionParams = opts.connectionParams;
+
+    const lazyOptions = {
+      ...lazyDefaults,
+      ...opts.lazy,
+    };
+    this.lazyEnabled = lazyOptions.enabled;
+
+    // Set up inactivity timeout for lazy connections.
+    this.inactivityTimeout = new ResettableTimeout(() => {
+      if (
+        this.requestManager.hasPendingRequests() ||
+        this.requestManager.hasActiveRequests()
+      ) {
+        this.inactivityTimeout.reset();
+        return;
+      }
+
+      void this.close();
+    }, lazyOptions.closeMs);
+
+    // Initialize the WebSocket connection.
+    this.activeConnection = new WsConnection({
+      WebSocketPonyfill: opts.WebSocket,
+      promiseUrl: prepareUrl(opts.url, !!opts.connectionParams).catch(
+        (error) => {
+          throw new TRPCWebSocketReconnectFatal({
+            message:
+              'Error when building url. Ensure provided url(): Promise<string> does not throws.',
+            cause: error,
+          });
```

Comment:
This should be redundant as you do it elsewhere

Reply:
Removed TRPCWebSocketReconnectFatal entirely. I initially used it to handle errors that I assumed couldn't be recovered from, but I realized this was overly complex. Now all connection errors are handled uniformly - they trigger a retry (if lazy is false) and notify through callbacks